kensu/utils/reporters.py

```diff
@@ -0,0 +1,165 @@
+        elif name == "LoggingReporter":
+            reporter = LoggingReporter(config)
+        elif name == "FileReporter":
+            reporter = FileReporter(config)
+        elif name == "ApiReporter":
+            reporter = ApiReporter(config)
+        elif name == "KafkaReporter":
+            reporter = KafkaReporter(config)
+        elif name == "MultiReporter":
+            reporter = MultiReporter(config)
+        return reporter
+
+
+class GenericReporter(Reporter):
+    def __init__(self, config, fun):
+        super().__init__(config)
+        self.fun = fun
+    def apply(self, obj, kensu_api, method):
+        return fun(obj, kensu_api, method)
+
+
+class DoNothingReporter(Reporter):
+    def apply(self, obj, kensu_api, method):
+        return obj
+
+
+class PrintReporter(Reporter):
+    def apply(self, obj, kensu_api, method):
+        json = self.entity_to_json_event(obj, kensu_api)
+        print(json)
+        return obj
+
+
+class LoggingReporter(Reporter):
+    def __init__(self, config, level=None):
+        super().__init__(config)
+        self.level = level
+        if self.level is None and config is not None:
+            self.level = config["level"]
+
+    def apply(self, obj, kensu_api, method):
+        json = self.entity_to_json_event(obj, kensu_api)
+        # TODO check variable and fun names :-D
+        if self.level.lower() == "info":
+            logging.info(json)
+        elif self.level.lower() == "warn":
+            logging.warn(json)
+        elif self.level.lower() == "error":
+            logging.error(json)
+        elif self.level.lower() == "debug":
+            logging.debug(json)
+        return obj
+
+
+class FileReporter(Reporter):
+    def __init__(self, config, file_name=None):
+        super().__init__(config)
+        if config is not None:
+            file_name = file_name or config["file_name"]
+        self.file = open(file_name, "a")
+        import atexit
+        atexit.register(self.close_gracefully)
+
+    def close(self):
+        if self.file is not None:
+            self.file.close()
+
+    def close_gracefully(self):
+        try:
+            self.close()
+        except IOError as ex:
+            print('Failed closing the output file', ex)
+
+    def apply(self, obj, kensu_api, method):
+        self.file.write(self.entity_to_json_event(obj, kensu_api))
+        self.file.flush()
+        return obj
+
+
+class ApiReporter(Reporter):
+    def apply(self, obj, kensu_api, method):
+        method(obj)
+        return obj
+
+
+class KafkaReporter(Reporter):
+    def __init__(self, config):
+        super().__init__(config)
+        if config is not None:
+            from confluent_kafka import Producer
+            import socket
+            self.bootstrap_servers = ",".join(json.loads(config["bootstrap_servers"]))
+            conf = {'bootstrap.servers': self.bootstrap_servers,
+                    'client.id': socket.gethostname()}
```

Comment:
what is this `client.id` used for? what if there are multiple producers on same hostname?

Reply:
not much of a problem apparently... seems to be interesting for logging/debugging https://jaceklaskowski.gitbooks.io/apache-kafka/content/kafka-properties-client-id.html